Index: acpica-unix2-20130517/source/common/dmtable.c
===================================================================
--- acpica-unix2-20130517.orig/source/common/dmtable.c
+++ acpica-unix2-20130517/source/common/dmtable.c
@@ -744,6 +744,10 @@ AcpiDmDumpTable (
             ByteLength = 8;
             break;
 
+        case ACPI_DMT_BUF10:
+	    ByteLength = 10;
+	    break;
+
         case ACPI_DMT_BUF16:
         case ACPI_DMT_UUID:
 
@@ -862,6 +866,7 @@ AcpiDmDumpTable (
             break;
 
         case ACPI_DMT_BUF7:
+        case ACPI_DMT_BUF10:
         case ACPI_DMT_BUF16:
         case ACPI_DMT_BUF128:
             /*
Index: acpica-unix2-20130517/source/common/dmtbinfo.c
===================================================================
--- acpica-unix2-20130517.orig/source/common/dmtbinfo.c
+++ acpica-unix2-20130517/source/common/dmtbinfo.c
@@ -780,6 +780,7 @@ ACPI_DMTABLE_INFO           AcpiDmTableI
 {
     {ACPI_DMT_UINT8,    ACPI_DMAR_OFFSET (Width),                   "Host Address Width", 0},
     {ACPI_DMT_UINT8,    ACPI_DMAR_OFFSET (Flags),                   "Flags", 0},
+    {ACPI_DMT_BUF10,    ACPI_DMAR_OFFSET (Reserved[0]),             "Reserved", 0},
     ACPI_DMT_TERMINATOR
 };
 
Index: acpica-unix2-20130517/source/compiler/dttable.c
===================================================================
--- acpica-unix2-20130517.orig/source/compiler/dttable.c
+++ acpica-unix2-20130517/source/compiler/dttable.c
@@ -541,8 +541,9 @@ DtCompileDmar (
     DT_FIELD                *SubtableStart;
     ACPI_DMTABLE_INFO       *InfoTable;
     ACPI_DMAR_HEADER        *DmarHeader;
-    UINT8                   *ReservedBuffer;
-    UINT32                  ReservedSize;
+    ACPI_DMAR_DEVICE_SCOPE  *DmarDeviceScope;
+    UINT32                  DeviceScopeLength;
+    UINT32                  PciPathLength;
 
 
     Status = DtCompileTable (PFieldList, AcpiDmTableInfoDmar, &Subtable, TRUE);
@@ -553,17 +554,7 @@ DtCompileDmar (
 
     ParentTable = DtPeekSubtable ();
     DtInsertSubtable (ParentTable, Subtable);
-
-    /* DMAR Reserved area */
-
-    ReservedSize = (UINT32) sizeof (((ACPI_TABLE_DMAR *) NULL)->Reserved);
-    ReservedBuffer = UtLocalCalloc (ReservedSize);
-
-    DtCreateSubtable (ReservedBuffer, ReservedSize, &Subtable);
-
-    ACPI_FREE (ReservedBuffer);
-    ParentTable = DtPeekSubtable ();
-    DtInsertSubtable (ParentTable, Subtable);
+    DtPushSubtable (Subtable);
 
     while (*PFieldList)
     {
@@ -621,10 +612,12 @@ DtCompileDmar (
 
         ParentTable = DtPeekSubtable ();
         DtInsertSubtable (ParentTable, Subtable);
+        DtPushSubtable (Subtable);
 
         /* Optional Device Scope subtables */
 
-        while (*PFieldList)
+        DeviceScopeLength = DmarHeader->Length - Subtable->Length - ParentTable->Length;
+        while (DeviceScopeLength)
         {
             Status = DtCompileTable (PFieldList, AcpiDmTableInfoDmarScope,
                         &Subtable, FALSE);
@@ -637,9 +630,12 @@ DtCompileDmar (
             DtInsertSubtable (ParentTable, Subtable);
             DtPushSubtable (Subtable);
 
+            DmarDeviceScope = ACPI_CAST_PTR (ACPI_DMAR_DEVICE_SCOPE, Subtable->Buffer);
+
             /* Optional PCI Paths */
 
-            while (*PFieldList)
+            PciPathLength = DmarDeviceScope->Length - Subtable->Length;
+            while (PciPathLength)
             {
                 Status = DtCompileTable (PFieldList, TableInfoDmarPciPath,
                             &Subtable, FALSE);
@@ -651,10 +647,15 @@ DtCompileDmar (
 
                 ParentTable = DtPeekSubtable ();
                 DtInsertSubtable (ParentTable, Subtable);
+                PciPathLength -= Subtable->Length;
             }
+
+            DtPopSubtable ();
+            DeviceScopeLength -= DmarDeviceScope->Length;
         }
 
         DtPopSubtable ();
+        DtPopSubtable ();
     }
 
     return (AE_OK);
Index: acpica-unix2-20130517/source/compiler/dtutils.c
===================================================================
--- acpica-unix2-20130517.orig/source/compiler/dtutils.c
+++ acpica-unix2-20130517/source/compiler/dtutils.c
@@ -397,6 +397,7 @@ DtGetFieldType (
 
     case ACPI_DMT_BUFFER:
     case ACPI_DMT_BUF7:
+    case ACPI_DMT_BUF10:
     case ACPI_DMT_BUF16:
     case ACPI_DMT_BUF128:
     case ACPI_DMT_PCI_PATH:
@@ -630,6 +631,10 @@ DtGetFieldLength (
         }
         break;
 
+    case ACPI_DMT_BUF10:
+        ByteLength = 10;
+        break;
+
     case ACPI_DMT_BUF16:
     case ACPI_DMT_UUID:
 
Index: acpica-unix2-20130517/source/include/acdisasm.h
===================================================================
--- acpica-unix2-20130517.orig/source/include/acdisasm.h
+++ acpica-unix2-20130517/source/include/acdisasm.h
@@ -108,6 +108,7 @@ typedef enum
     ACPI_DMT_UINT56,
     ACPI_DMT_UINT64,
     ACPI_DMT_BUF7,
+    ACPI_DMT_BUF10,
     ACPI_DMT_BUF16,
     ACPI_DMT_BUF128,
     ACPI_DMT_SIG,
